pkg/controllers/dnspolicy/dnspolicy_dnsrecords.go

```diff
@@ -79,42 +124,269 @@ func (r *DNSPolicyReconciler) reconcileGatewayDNSRecords(ctx context.Context, ga
-			log.V(3).Info("hostHasAttachedRoutes", "host", mh.Host, "hostHasAttachedRoutes", attached)
-			cg, err := r.Placement.GetClusterGateway(ctx, gateway, downstreamCluster)
+			cg, err := r.Placement.GetClusterGateway(ctx, upstreamGateway, downstreamCluster)
 			if err != nil {
 				return fmt.Errorf("get cluster gateway failed: %s", err)
 			}
 			clusterGateways = append(clusterGateways, cg)
-		}
-
-		if len(clusterGateways) == 0 {
-			// delete record
-			log.V(3).Info("no cluster gateways, deleting DNS record", " for host ", mh.Host)
-			if mh.DnsRecord != nil {
-				if err := r.Client().Delete(ctx, mh.DnsRecord); client.IgnoreNotFound(err) != nil {
-					return fmt.Errorf("failed to deleted dns record for managed host %s : %s", mh.Host, err)
+			if len(clusterGateways) == 0 {
+				// delete record
+				log.V(3).Info("no cluster gateways, deleting DNS record", " for gateway ", upstreamGateway.Name, "and listener", listener.Name)
+				if err := r.deleteDNSRecordForListener(upstreamGateway.Namespace, listener); client.IgnoreNotFound(err) != nil {
+					return fmt.Errorf("failed to deleted dns record for managed host %s : %s", listener.Name, err)
 				}
+				return nil
 			}
-			return nil
 		}
-		var dnsRecord, err = r.HostService.CreateDNSRecord(ctx, mh.Subdomain, mh.ManagedZone, gateway)
+		dnsRecord, err := r.createListenerDNSRecord(ctx, dnsPolicy.GetObjectMeta().GetUID(), upstreamGateway, mz, listener)
 		if err := client.IgnoreAlreadyExists(err); err != nil {
-			return fmt.Errorf("failed to create dns record for host %s : %s ", mh.Host, err)
+			return fmt.Errorf("failed to create dns record for listener %s : %s ", listener.Name, err)
 		}
 		if k8serrors.IsAlreadyExists(err) {
-			dnsRecord, err = r.HostService.GetDNSRecord(ctx, mh.Subdomain, mh.ManagedZone, gateway)
-			if err != nil {
-				return fmt.Errorf("failed to get dns record for host %s : %s ", mh.Host, err)
+			if err := r.Client().Get(ctx, client.ObjectKeyFromObject(dnsRecord), dnsRecord); err != nil {
+				return fmt.Errorf("failed to get dns record for listener %s : %s ", listener.Name, err)
 			}
 		}
 
-		mcgTarget := dns.NewMultiClusterGatewayTarget(gateway, clusterGateways, dnsPolicy.Spec.LoadBalancing)
-		log.Info("setting dns dnsTargets for gateway listener", "listener", dnsRecord.Name, "values", mcgTarget)
+		mcgTarget := dns.NewMultiClusterGatewayTarget(upstreamGateway, clusterGateways, dnsPolicy.Spec.LoadBalancing)
+		log.Info("setting dns dnsTargets for gateway listener", "listener", listener.Name, "values", mcgTarget)
 
-		if err := r.HostService.SetEndpoints(ctx, mcgTarget, dnsRecord); err != nil {
+		if err := r.SetEndpoints(ctx, mcgTarget, dnsRecord, &listener); err != nil {
 			return fmt.Errorf("failed to add dns record dnsTargets %s %v", err, mcgTarget)
 		}
+	}
+	return nil
+}
+
+func (r *DNSPolicyReconciler) SetEndpoints(ctx context.Context, mcgTarget *dns.MultiClusterGatewayTarget, dnsRecord *v1alpha1.DNSRecord, listener *gatewayv1beta1.Listener) error {
```

Comment:
This can probably be private. This is also out of date with what is on main. I wonder if you are moving all these things from the service if we shouldn't do that in it's own PR without functional changes. We'd be able to get it merged quicker.

Reply:
ok will spit some of this out